src/txn/txn_timestamp.c

```diff
@@ -691,13 +691,14 @@ __wt_txn_parse_read_timestamp(WT_SESSION_IMPL *session, const char *cfg[])
 		txn_global = &S2C(session)->txn_global;
 		WT_RET(__wt_txn_parse_timestamp(session, "read", &ts, &cval));
 
-		/* Read timestamps imply / require snapshot isolation. */
-		if (!F_ISSET(txn, WT_TXN_RUNNING))
-			txn->isolation = WT_ISO_SNAPSHOT;
-		else if (txn->isolation != WT_ISO_SNAPSHOT)
-			WT_RET_MSG(session, EINVAL, "setting a read_timestamp"
-			    " requires a transaction running at snapshot"
-			    " isolation");
+		/*
+		 * Read timestamps imply / require snapshot isolation.
+		 *
+		 * If we already have a snapshot, it may be too early to match
+		 * the timestamp.  Get a new one.
+		 */
+		txn->isolation = WT_ISO_SNAPSHOT;
```

Comment:
Changing the isolation level doesn't feel safe to me. Could this be changed to an error if it isn't already `WT_ISO_SNAPSHOT`? From what I can tell the isolation level will always have been setup by the time we get here.

Reply:
OK, I've reverted that part of the change.  I think we may want to reconsider in WT-4000, but let's keep this change simple.